### DISCUSSION TITLE
Why this happens?

### DISCUSSION THREAD

**Discussion by vega0** [antlr/antlr4#4268] [General]
https://www.youtube.com/watch?v=rfr-CSKrues&feature=youtu.be

In single file it works fine except single character numbers, but after combining into one file it enexpective behave into errors that i can't understand. Why its happens. 

Repo : https://gitflic.ru/project/nightmare0/rossijskij-yap

**Comment by Marti2203** [comment]
I suggest debugging by seeing the emitted tokens to get the better idea as the issue is not on the parser level but the lexer level - the lexer tries to give the longest possible string to a rule, no matter the order so if you have xstringa it will return an identifier 'xstringa' and not 'x' 'string' 'a' tokens.  The other issue is that you define the digits as speical tokens which will cause some problems. I suggest removing the W_НОМЕР_НОЛЬ and inlining the values.

**Comment by kaby76** [comment]
The Eclipse IDE does not show the token type, which is again why you need to stop using it.

Please read this carefully: ***An Antlr4 lexer works independently from the parser. The lexer matches the LONGEST string first regardless of the order of the lexer rules. But, if two or more lexer rules match a string of the same length, the first one listed is selected.***

"E+3456" is matched by `FRAGMENTPUTI`. `aЕпсилонЧастьЧисла` never matches because the parser does not guide how to lex. The lexer reads all the input and tokenizes the input before the parser does anything. As you type in the input string in Eclipse, the first thing that is done is to tokenize the entire string from the first character to the last. The parser is not even involved. `ZNAK_EPSILON` cannot possibly ever match. The lexer will always choose the longest string. "E+3456" (6 characters long) is longer than "E" (one character long).

**Comment by vega0** [comment]
the problem means separately recogninition except full recognition ?

**Reply by kaby76** [reply]
Yes, the lexer and parser are separate recognizers. The lexer creates a list of tokens independently from the parser using just the lexer rules from YapComplete.g4. Then, after all input is read and the list of tokens created, it is inputted to the parser, which is defined by the parser rules in YapComplete.g4.

**Reply by vega0** [reply]
how to test tokens?

**Reply by vega0** [reply]
how to separate import?

**Reply by vega0** [reply]
best rule of import.

**Reply by kaby76** [reply]
> how to test tokens?

Stop using Eclipse, and use Intellij.

* Open YapComplete.g4 using Intellij https://www.jetbrains.com/idea/.
* Add in Antlr4 Plugin.
    * Settings. ![Screenshot (53)](https://github.com/antlr/antlr4/assets/11389451/a30023dc-2fee-41cb-8673-f6756025bb83)
    * Plugins. ![Screenshot (54)](https://github.com/antlr/antlr4/assets/11389451/dd2c6bb5-538d-4eaa-adbd-38803a0ebcf7)
    * Search marketplace for "antlr4".  ![Screenshot (55)](https://github.com/antlr/antlr4/assets/11389451/33b68af8-8526-4160-a866-ef5d82f275f7)
    * Install and restart. ![Screenshot (56)](https://github.com/antlr/antlr4/assets/11389451/149b966e-885b-44a3-b913-6a76ece11a75)
* Configure Antlr4. ![Screenshot (47)](https://github.com/antlr/antlr4/assets/11389451/3aa3eebb-853d-4a46-a8c7-3efbd15658ba)
* YapComplete.g4 is in UTF8. Set UTF8. ![Screenshot (48)](https://github.com/antlr/antlr4/assets/11389451/f8442cef-a04d-4eb2-bc7c-21804ae7aa7e)
* Generate ANTLR Recognizer. ![Screenshot (49)](https://github.com/antlr/antlr4/assets/11389451/284b99e3-491a-4a39-afd3-b6389873e163)
* Verify generate. ![Screenshot (50)](https://github.com/antlr/antlr4/assets/11389451/e7f39046-7bb9-48ef-9e1b-50592f0917ed)
* Right-click on parser rule, test rule. ![Screenshot (51)](https://github.com/antlr/antlr4/assets/11389451/d03fd2e6-4947-4905-b7fb-a0e83d3490d7)
* Type input string, see output with token type and token value. ![Screenshot (52)](https://github.com/antlr/antlr4/assets/11389451/a8ac4fa3-4e65-4368-95de-087672f471de)

**Reply by vega0** [reply]
https://gitflic.ru/project/nightmare0/rossijskij-yap did all that you told me and got wrong parser behavior. Separately it works fine but in assembled file its throws recognition error.  No errors when building antlr4 data. 

[image](https://i.imgur.com/RRaZV7w.png)

**Reply by vega0** [reply]
uhm, explain https://i.imgur.com/vndkQ0v.png this token recognition mistake that
 makes no sense 
 
 
```
 // ФАЙЛ ОБЪХЕКТА StrokaFailovyPut
fragmentputi: FRAGMENTPUTI rashireniyeFaila?;
rashireniyeFaila: TOCHKA fragmentputi;
failovyPut: '@' '/'? fragmentputi ('/' fragmentputi)*;
```

there is no zero defined chars that can be triggered by token...

**Comment by kaby76** [comment]
Yes, I can reproduce that. ![Screenshot (57)](https://github.com/antlr/antlr4/assets/11389451/8625e0ff-90e1-49a9-b545-3195f11c02b3)

That is *exactly* how Antlr is supposed to work. ***An Antlr4 lexer works independently from the parser. The lexer matches the LONGEST string first regardless of the order of the lexer rules. But, if two or more lexer rules match a string of the same length, the first one listed is selected.***

In YapComplete.g4, you have these two rules:
```
A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ: (A_ДЕСЯТЕРИЧНОЕ_ЧИСЛО | ZNAKI_A_F)+;
...
FRAGMENTPUTI: ('\\ ' | ~(' ' | '/' | '.'))+;
```
"abc" matches both A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ and FRAGMENTPUTI. Both matches are three characters in length. But, A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ (line 95) occurs before FRAGMENTPUTI (line 99) in YapComplete.g4. Therefore A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ "wins". It does not matter that it is trying to match parser rule fragmentputi. The parser never "guides" how the lexer to do its job. The lexer works independently from the parser. The lexer gathers all lexer rules from YapComplete.g4, and tries to match each rule from top to bottom in order.

If you don't believe me, rewrite A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ  to something like `A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ  : 'aaaaaaaa';`. Then, A_ШЕСТНАДЦАТИРИЧНЫЕ_СИМВОЛЫ  does not compete with FRAGMENTPUTI.
![Screenshot (58)](https://github.com/antlr/antlr4/assets/11389451/926cb3b6-204f-4018-bed7-6f2101ded080)

**Reply by vega0** [reply]
is it possible to split lexer?
